cpp/meta_module/algorithm/meta.cpp

```diff
@@ -0,0 +1,284 @@
+#include "meta.hpp"
+#include <cstdint>
+#include <string>
+#include <string_view>
+#include "mgp.hpp"
+
+namespace Meta {
+
+namespace {
+
+class Metadata {
+ public:
+  Metadata() { node_cnt = relationship_cnt = 0; }
+
+  int64_t node_cnt;
+  int64_t relationship_cnt;
+  std::unordered_map<std::string, int64_t> labels;
+  std::unordered_map<std::string, int64_t> property_key_cnt;
+  std::unordered_map<std::string, int64_t> relationship_types;
+  std::unordered_map<std::string, int64_t> relationship_types_cnt;
+
+  int64_t get_label_count() const { return static_cast<int64_t>(labels.size()); }
+  int64_t get_relationship_type_count() const { return static_cast<int64_t>(relationship_types_cnt.size()); }
+  int64_t get_property_key_count() const { return static_cast<int64_t>(property_key_cnt.size()); }
+
+  void reset();
+  void update_labels(const mgp::Node &node, int add);
+  void update_property_key_cnt(const mgp::Node &node, int add);
+  void update_property_key_cnt(const mgp::Relationship &relationship, int add);
+  void update_relationship_types(const mgp::Relationship &relationship, int add);
+  void update_relationship_types_cnt(const mgp::Relationship &relationship, int add);
+};
+
+// Global variable
+// This is unsafe in the multithreaded environment, the workaround would be building a thread-safe dynamic storage
+// implementation
+Metadata metadata;
```

Comment:
I think you are right, but we can do one other thing in my opinion. We can have lock per variable, and in that case when you want to access `labels` you take `lables` lock. You update labels and that should be all right. What will we do with that? I think we will have thread safe, but not "fully correct" implementation. Because if one thread updates labels, and other updates properties, it might happen that properties thread will get update of labels, but it didn't make such change. And that is all right, because it actually takes into account commits, and it works as is. I think it is more granular this way than using one lock for whole object. Also, for counters `relationship_cnt` and `node_cnt` we can use atomics.

Reply:
@Josipmrden @antepusic tagging you here too